channeldb/messagestore.go

```diff
@@ -0,0 +1,227 @@
+package channeldb
+
+import (
+	"bytes"
+	"encoding/binary"
+	"sort"
+
+	"github.com/boltdb/bolt"
+	"github.com/go-errors/errors"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+var (
+	// basePeerBucketKey is the base key for generating peer bucket keys.
+	basePeerBucketKey = []byte("peermessages")
+)
+
+// MessagesStore represents the boltdb storage for messages inside
+// retransmission sub-system. For every peer we have number of buckets which
+// represent the different type of messages, such logic is due to the fact
+// that by receiving ack messages we need to delete all messages of specific
+// types.
+type MessagesStore struct {
+	// id is a unique identificator of peer.
+	id []byte
+	db *DB
+
+	// cache stores the correspondence of message code and bool which
+	// denotes do messages of this type are exist in storage.
+	cache map[lnwire.Code]bool
+}
+
+// NewMessageStore creates new instance of message storage.
+func NewMessageStore(id []byte, db *DB) *MessagesStore {
+	cache := make(map[lnwire.Code]bool)
+
+	// Initially initialize message cache like if we have all types of
+	// messages.
+	for _, code := range lnwire.AllCommands() {
+		cache[code] = true
+	}
+
+	return &MessagesStore{
+		id:    id,
+		db:    db,
+		cache: cache,
+	}
+}
+
+// Add add message to the storage.
+func (s *MessagesStore) Add(msg lnwire.Message) error {
+	if err := s.db.Batch(func(tx *bolt.Tx) error {
+		peerBucketKey := s.getPeerBucketKey()
+		peerBucket, err := tx.CreateBucketIfNotExists(peerBucketKey)
+		if err != nil {
+			return err
+		}
+
+		// We use integer representation of code as bucket key due to
+		// the fact that lately we will need to restore messages. And
+		// such approach allows as not to do additional steps for
+		// determine the message type.
+		code := msg.Command()
+		key := make([]byte, 4)
+		binary.BigEndian.PutUint32(key, uint32(code))
+
+		messageBucket, err := peerBucket.CreateBucketIfNotExists(key)
+		if err != nil {
+			return err
+		}
+
+		var b bytes.Buffer
+		if err := msg.Encode(&b, 0); err != nil {
+			return err
+		}
+
+		// Generate next sequence number to preserver the message order.
+		sequence, err := peerBucket.NextSequence()
```

Comment:
Switching to the schema above eliminates the unnecessary sorting logic and also still retains the message order required to properly perform retransmissions.